aea/helpers/async_utils.py

```diff
@@ -0,0 +1,324 @@
+    """
+
+    def __init__(self, coro: Coroutine, loop: AbstractEventLoop) -> None:
+        """
+        Init the task.
+
+        :param coro: coroutine to schedule
+        :param loop: an event loop to schedule on.
+        """
+        self._task = loop.create_task(coro)
+        self._loop = loop
+        self._future = asyncio.run_coroutine_threadsafe(self._get_task_result(), loop)
+
+    async def _get_task_result(self) -> Any:
+        """Get task result, should be run in target loop."""
+        return await self._task
+
+    def result(self, timeout: Optional[float] = None) -> Any:
+        """
+        Wait for coroutine execution result.
+
+        :param timeout: optional timeout to wait in seconds.
+        """
+        return self._future.result(timeout)
+
+    def cancel(self) -> None:
+        """Cancel coroutine task execution in a target loop."""
+        self._loop.call_soon_threadsafe(self._task.cancel)
+
+    def future_cancel(self) -> None:
+        """
+        Cancel task waiting future.
+
+        In this case future result will raise CanclledError not waiting for real task exit.
+        """
+        self._future.cancel()
+
+    def done(self) -> bool:
+        """Check task is done."""
+        return self._future.done()
+
+
+class ThreadedAsyncRunner(Thread):
+    """Util to run thread with event loop and execute coroutines inside."""
+
+    def __init__(self, loop=None) -> None:
+        """
+        Init threaded runner.
+
+        :param loop: optional event loop. is it's running loop, threaded runner will use it.
+        """
+        self._loop = loop or asyncio.new_event_loop()
+        assert not self._loop.is_closed()
+        super().__init__(daemon=True)
+
+    def start(self) -> None:
+        """Start event loop in dedicated thread."""
+        if self.is_alive() or self._loop.is_running():
+            return
+        super().start()
+        self.call(asyncio.sleep(0.001)).result(1)  # type: ignore # to ensure loop is started
+
+    def run(self) -> None:
+        """Run code inside thread."""
+        logger.debug("Starting threaded asyncio loop...")
+        asyncio.set_event_loop(self._loop)
+        self._loop.run_forever()
+        logger.debug("Asyncio loop has been stopped.")
+
+    def call(self, coro: Coroutine) -> Any:
+        """
+        Run a coroutine inside the event loop.
+
+        :param coro: a coroutine to run.
+        """
+        return AnotherThreadTask(coro, self._loop)
+
+    def stop(self) -> None:
+        """Stop event loop in thread."""
+        logger.debug("Stopping...")
+        if not self.is_alive():
+            return
+        if self._loop.is_running():
+            logger.debug("Stopping loop...")
+            self._loop.call_soon_threadsafe(lambda: self._loop.stop())
+        logger.debug("Wait thread to join...")
+        self.join(10)
+        logger.debug("Stopped.")
+
+
+async def cancel_and_wait(task: Optional[Task]) -> Any:
```

Comment:
We have `wait_and_cancel` and `cancel_and_wait`. It would help if these functions where next to each other so the reader spots the difference immediately.